pgmpy/models/BayesianNetwork.py

```diff
@@ -753,8 +755,45 @@ def predict(self, data, algo=None, stochastic=False, n_jobs=-1, **kwargs):
                 )
             )
 
+            # example 'data' used for prediction
+            #         A    B    C    D
+            #   800  1.0  NaN  0.0  NaN
+            #   801  1.0  0.0  NaN  0.0
+
             df_results = pd.DataFrame(pred_values, index=data_unique.index)
-            data_with_results = pd.concat([data_unique, df_results], axis=1)
+            print(df_results)
+            # converting results dict to dataframe, only the predicted data points are not NaN
+            #        E    B    D    C    A
+            #  800  1.0  1.0  1.0  NaN  NaN
+            #  801  0.0  NaN  NaN  1.0  NaN
+
+            all_columns = data_unique.columns.tolist() + [
+                col for col in df_results.columns if col not in data_unique.columns
+            ]
+            df_complete_results = df_results.reindex(columns=all_columns)
+            print(df_complete_results)
```

Comment:
I can't think of an easy way to get the final dataframe using merge when the Nan values are present. But you can do something similar to the logic used in the `stochastic=True` case: https://github.com/pgmpy/pgmpy/blob/dev/pgmpy/models/BayesianNetwork.py#L715. This line creates a mapping of duplicate columns and then utilizes this mapping to create the final results data frame. The main difference in this case would be that we won't do sampling as done at: https://github.com/pgmpy/pgmpy/blob/dev/pgmpy/models/BayesianNetwork.py#L733 but rather it would be the same value for each duplicated row.